soc: andestech: add support for ae350/clic SoC

jimmyzhe opened the pull request:
Adds support for ae350/clic SoC and the target `adp_xc7k/ae350/clic`.

The AE350 CLIC shares the same peripherals as AE350 PLIC, with the following differences:
1. Use CLIC for interrupt controller instead of PLIC
2. Different interrupt source number
All interrupts, including local interrupts (MTIP/MSIP), are controlled by CLIC.
3. Supports only one CPU


This PR depends on 
- #85905

jimmyzhe committed:
dts: riscv: andes: add andes_v5_ae350_clic.dtsi

Add andes_v5_ae350_clic for the AE350 FPGA platform with CLIC support.

Signed-off-by: Jimmy Zheng <jimmyzhe@andestech.com>

jimmyzhe committed:
soc: andestech: ae350: add ae350/clic configuration

Add ae350/clic soc, which shares the same peripherials as AE350 PLIC
platform but uses CLIC instead of PLIC, with different IRQ number.

Signed-off-by: Jimmy Zheng <jimmyzhe@andestech.com>

jimmyzhe committed:
boards: andestech: add adp_xc7k/ae350/clic target

Add adp_xc7k/ae350/clic target and saperate the ae350 common DTS into
adp_xc7k_ae350_common.dtsi.

Signed-off-by: Jimmy Zheng <jimmyzhe@andestech.com>

jimmyzhe committed:
tests: kernel: interrupt: adjust test for clic

For RISC-V CLIC, only edge-triggered interrupts can be triggered by
software. Add IRQ flag to set the trigger type for RISC-V CLIC.

Signed-off-by: Jimmy Zheng <jimmyzhe@andestech.com>

jimmyzhe committed:
tests: kernel: gen_isr_table: add available IRQ number for AE350 CLIC

RISC-V CLIC can trigger edge-triggered interrupts by software, but the
available IRQ sources depend on the hardware implementation.
Clarifies the IRQ source for GD32VF103 and adds support for AE350 CLIC.

Signed-off-by: Jimmy Zheng <jimmyzhe@andestech.com>

jimmyzhe committed:
tests: drivers: add overlay files for adp_xc7k/ae350/clic

adp_xc7k/ae350/clic uses a different interrupt controller than
adp_xc7k/ae350, add overlay files as in adp_xc7k/ae350.

Signed-off-by: Jimmy Zheng <jimmyzhe@andestech.com>

jimmyzhe committed:
samples: drivers: watchdog: add overlay file for adp_xc7k/ae350/clic

adp_xc7k/ae350/clic uses a different interrupt controller than
adp_xc7k/ae350, add overlay files as in adp_xc7k/ae350.

Signed-off-by: Jimmy Zheng <jimmyzhe@andestech.com>

jimmyzhe commented:
Rebases to main.

Two kernel test cases are modified in this PR because:

1. **tests/kernel/interrupt**
RISC-V CLIC requires edge-triggered configuration for software-triggered interrupts. Added IRQ_FLAGS as already done in [tests/kernel/gen_isr_table](https://github.com/zephyrproject-rtos/zephyr/blob/fd4cc0e0310d1c4d9a423bad9a6fa65cd2dcbc8d/tests/kernel/gen_isr_table/src/main.c#L121).

2. **tests/kernel/gen_isr_table**
Adjusted the tested IRQ number for adp_xc7k/ae350/clic.

fkokosinski reviewed (dismissed):
LGTM

cfriedt commented on boards/andestech/adp_xc7k_ae350/adp_xc7k_ae350_clic_defconfig:
```diff
@@ -0,0 +1,47 @@
+CONFIG_HWINFO=y
+CONFIG_HWINFO_ANDES=y
+
+# RV64 options
+#CONFIG_RV64I_CPU=y
+
+# FPU (single precision) options
+#CONFIG_FPU=y
+#CONFIG_FPU_SHARING=y
+#CONFIG_FLOAT_HARD=y
+#CONFIG_SINGLE_PRECISION_FPU=y
+
+# FPU (double precision) options
+#CONFIG_FPU=y
+#CONFIG_FPU_SHARING=y
+#CONFIG_FLOAT_HARD=y
+#CONFIG_DOUBLE_PRECISION_FPU=y
+
+# PMP granularity options
+CONFIG_PMP_GRANULARITY=8
+
+# HW DSP options
+#CONFIG_SOC_ANDES_V5_HWDSP=y
+
+# EXEC.IT options
+#CONFIG_SOC_ANDES_V5_EXECIT=y
+
+# Nocache memory options
+#CONFIG_SOC_ANDES_V5_PMA=y
+#CONFIG_NOCACHE_MEMORY=y
+
+# I/O Coherence Port options
+#CONFIG_SOC_ANDES_V5_IOCP=y
```
Just wanted to point out that normally we try to avoid commenting-out Kconfig options like this, but I don't think it's a blocking issue.

cfriedt reviewed (dismissed):
I suspect that this could potentially be rebased, and if so, I'll leave it up to one of the release managers. If there is a rebase, it would be nice to remove the commented-out Kconfig values.

Otherwise, LGTM

jimmyzhe committed:
boards: andestech: clean up default Kconfig for adp_xc7k/ae350

Remove redundant or commented-out Kconfig, and only enable the baseline
features compatible with most AE350 platforms.
Platform-specific extensions and features should be enabled by menuconfig.

Signed-off-by: Jimmy Zheng <jimmyzhe@andestech.com>

ycsin reviewed (dismissed):
The copyright year of the newly added files should be 2025 🤷‍♂️

jimmyzhe commented:
Rebased.
